File: Laba4/Java/ArbitratorStrategy.java
import java.util.concurrent.Semaphore;

class ArbitratorStrategy implements Strategy {
    private final Semaphore thinkingPhilosophers = new Semaphore(4, true);

    @Override
    public void takeForks(int id, Table table) {
        try {
            thinkingPhilosophers.acquire();
            table.getFork(id);
            table.getFork((id + 1) % Main.PHILOSOPHERS_COUNT);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    @Override
    public void putForks(int id, Table table) {
        table.putFork((id + 1) % Main.PHILOSOPHERS_COUNT);
        table.putFork(id);
        thinkingPhilosophers.release();
    }
}

File: Laba4/Sharp/ArbitratorStrategy.cs
namespace Laba4
{
    public class ArbitratorStrategy : IStrategy
    {
        private readonly Semaphore thinkingPhilosophers = new(4, 4);

        public void TakeForks(int id, Table table)
        {
            thinkingPhilosophers.WaitOne();
            table.GetFork(id);
            table.GetFork((id + 1) % Program.PhilosophersCount);
        }

        public void PutForks(int id, Table table)
        {
            table.PutFork((id + 1) % Program.PhilosophersCount);
            table.PutFork(id);
            thinkingPhilosophers.Release();
        }
    }
}